crates/bevy_ecs/src/archetype.rs

```diff
@@ -124,47 +126,91 @@ pub struct Archetype {
     entities: Vec<Entity>,
     edges: Edges,
     table_info: TableInfo,
-    table_components: Cow<'static, [ComponentId]>,
-    sparse_set_components: Cow<'static, [ComponentId]>,
-    pub(crate) unique_components: SparseSet<ComponentId, Column>,
-    pub(crate) components: SparseSet<ComponentId, ArchetypeComponentInfo>,
+    table_components: Cow<'static, [(RelationKindId, Option<Entity>)]>,
+    sparse_set_components: Cow<'static, [(RelationKindId, Option<Entity>)]>,
+    pub(crate) unique_components: SparseSet<RelationKindId, Column>,
+    pub(crate) components: SparseSet<
+        RelationKindId,
+        (
+            Option<ArchetypeComponentInfo>,
```

Comment:
`ArchetypeComponentInfo` already has a niche because the `StorageType` enum inside it only has two variants.